File: 1. 단계별로_풀어보기/10~19/16. 스택, 큐, 덱 1/5. 12789.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const numArr = inputArr[1].split(' ').map(Number)

// 추가 대기열 확보
const extraStack = Array.from({length: 1000}).fill(0)
let extraIdx = 0
let numIdx = 0

// 결과 출력 변수
let result = 'Nice'

// 1번부터 연산
for (let i = 1; i <= N; i++) {
  let isFind = false

  // 추가 대기열 맨 앞에 있는지 먼저 확인
  if (extraIdx > 0 && extraStack[extraIdx - 1] === i) {
    isFind = true
    extraIdx -= 1
    continue
  }

  // 원래 대기열에서 찾을 수 있나 확인
  for (numIdx = numIdx; numIdx < N; numIdx++) {
    if (numArr[numIdx] === i) {
      // 찾았으면 목록에서 제거한다.
      isFind = true
      numIdx += 1
      break
    } //
    else {
      // 못 찾았으면 추가 대기열로 넘긴다.
      extraStack[extraIdx] = numArr[numIdx]
      extraIdx += 1
    }
  }

  if (!isFind) {
    result = 'Sad'
    break
  }
}

process.stdout.write(result)

File: 1. 단계별로_풀어보기/10~19/16. 스택, 큐, 덱 1/5. 12789.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())
numArr = list(map(int, input().split()))
numIdx = 0

extraStack = deque()
result = 'Nice'

for i in range(1, N + 1):
  isFind = False

  if len(extraStack) > 0 and extraStack[-1] == i:
    isFind = True         # 불필요하긴 하다
    extraStack.pop()
    continue

  while numIdx < N:
    if numArr[numIdx] == i:
      numIdx += 1
      isFind = True
      break
    else:
      extraStack.append(numArr[numIdx])
    numIdx += 1

  if not isFind:
    result = 'Sad'
    break

output(f"{result}")
